src/rgw/rgw_iam_policy.cc

```diff
@@ -1269,6 +1271,18 @@ Effect Statement::eval_principal(const Environment& e,
     } else if (!noprinc.empty() && is_identity(*ida, noprinc)) {
       return Effect::Deny;
     }
+    if (ida->get_identity_type() == TYPE_RGW && !princ.empty()) {
+      bool princ_matched = false;
+      for (auto p : princ) {
+        if (ida->is_identity(p)) {
```

Comment:
doesn't this just duplicate the `is_identity()` call on line 1253 above?
```c++
    if (ida->get_identity_type() != TYPE_ROLE && !princ.empty() && !is_identity(*ida, princ)) {
```

Reply:
it is pretty similar. i can merge the code.